flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp

```diff
@@ -288,20 +294,51 @@ class GenericLoopConversionPattern
     rewriter.createBlock(&distributeOp.getRegion());
 
     mlir::omp::WsloopOperands wsloopClauseOps;
+    populateReductionClauseOps(loopOp, wsloopClauseOps);
+    Fortran::common::openmp::EntryBlockArgs wsloopArgs;
+    wsloopArgs.reduction.vars = wsloopClauseOps.reductionVars;
+
     auto wsloopOp =
         rewriter.create<mlir::omp::WsloopOp>(loopOp.getLoc(), wsloopClauseOps);
     wsloopOp.setComposite(true);
-    rewriter.createBlock(&wsloopOp.getRegion());
+    mlir::Block *loopBlock =
+        genEntryBlock(rewriter, wsloopArgs, wsloopOp.getRegion());
 
     mlir::IRMapping mapper;
-    mlir::Block &loopBlock = *loopOp.getRegion().begin();
 
-    for (auto [loopOpArg, parallelOpArg] : llvm::zip_equal(
-             loopBlock.getArguments(), parallelBlock->getArguments()))
+    auto loopBlockInterface =
+        llvm::cast<mlir::omp::BlockArgOpenMPOpInterface>(*loopOp);
+
+    for (auto [loopOpArg, parallelOpArg] :
+         llvm::zip_equal(loopBlockInterface.getPrivateBlockArgs(),
+                         parallelBlock->getArguments()))
       mapper.map(loopOpArg, parallelOpArg);
 
+    for (auto [loopOpArg, wsloopOpArg] :
+         llvm::zip_equal(loopBlockInterface.getReductionBlockArgs(),
+                         loopBlock->getArguments()))
+      mapper.map(loopOpArg, wsloopOpArg);
+
     rewriter.clone(*loopOp.begin(), mapper);
   }
+
+  template <typename OpOperandsTy>
+  void populateReductionClauseOps(mlir::omp::LoopOp loopOp,
+                                  OpOperandsTy &clauseOps) const {
```

Comment:
Nit: No need for templates here. Operation-associated clause operand structures already subclass the structure associated to each of their clauses ([link](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/#operand-structures)).
```suggestion
  void populateReductionClauseOps(mlir::omp::LoopOp loopOp,
                                  mlir::omp::ReductionClauseOps &clauseOps) const {
```